File: Android/dokit/src/main/java/com/didichuxing/doraemonkit/kit/colorpick/ScreenRecorderService.java
package com.didichuxing.doraemonkit.kit.colorpick;

import android.app.Activity;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.BitmapFactory;
import android.media.projection.MediaProjection;
import android.media.projection.MediaProjectionManager;
import android.os.Build;
import android.os.IBinder;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.didichuxing.doraemonkit.R;
import com.didichuxing.doraemonkit.kit.core.UniversalActivity;

/**
 * ================================================
 * 作    者：jint（金台）
 * 版    本：1.0
 * 创建日期：2020-02-21-17:36
 * 描    述：
 * 修订历史：
 * ================================================
 *
 */
public class ScreenRecorderService extends Service {

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        try {
            createNotificationChannel();
            //Android Q 存在兼容性问题
            MediaProjectionManager mMediaProjectionManager = (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
            MediaProjection mediaProjection = mMediaProjectionManager.getMediaProjection(Activity.RESULT_OK, (Intent) intent.getParcelableExtra("data"));
            ColorPickManager.getInstance().setMediaProjection(mediaProjection);
            if (ColorPickManager.getInstance().getColorPickerDokitView() != null) {
                ColorPickManager.getInstance().getColorPickerDokitView().onScreenServiceReady();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return super.onStartCommand(intent, flags, startId);
    }

    private void createNotificationChannel() {
        Notification.Builder builder = new Notification.Builder(this.getApplicationContext()); //获取一个Notification构造器
        Intent nfIntent = new Intent(this, UniversalActivity.class); //点击后跳转的界面，可以设置跳转数据
        PendingIntent pendingIntent;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            pendingIntent = PendingIntent.getActivity(this, 0, nfIntent, 0 | PendingIntent.FLAG_IMMUTABLE);
        } else {
            pendingIntent = PendingIntent.getActivity(this, 0, nfIntent, 0);
        }

        builder.setContentIntent(pendingIntent) // 设置PendingIntent
                .setLargeIcon(BitmapFactory.decodeResource(this.getResources(), R.mipmap.dk_doraemon)) // 设置下拉列表中的图标(大图标)
                //.setContentTitle("SMI InstantView") // 设置下拉列表里的标题
                .setSmallIcon(R.mipmap.dk_doraemon) // 设置状态栏内的小图标
                .setContentText("Dokit屏幕取色器前台服务") // 设置上下文内容
                .setWhen(System.currentTimeMillis()); // 设置该通知发生的时间

        /*以下是对Android 8.0的适配*/
        //普通notification适配
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            builder.setChannelId("notification_id");
        }
        //前台服务notification适配
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationManager notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
            NotificationChannel channel = new NotificationChannel("notification_id", "notification_name", NotificationManager.IMPORTANCE_LOW);
            notificationManager.createNotificationChannel(channel);
        }

        Notification notification = builder.build(); // 获取构建好的Notification
        notification.defaults = Notification.DEFAULT_SOUND; //设置为默认的声音
        startForeground(110, notification);

    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/colorpick/ScreenRecorderService.kt
package com.didichuxing.doraemonkit.kit.colorpick

import android.app.*
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.graphics.BitmapFactory
import android.os.Binder
import android.os.Build
import android.os.IBinder
import com.didichuxing.doraemonkit.R
import com.didichuxing.doraemonkit.kit.core.UniversalActivity

/**
 * 适配Android o。
 * 如果targetSdkVersion >=29 并且当前系统版本>= 26的时候必须设置前台service截屏。
 * @author Donald Yan
 * @date 2020/6/16
 */
class ScreenRecorderService : Service(){

    override fun onCreate() {
        super.onCreate()
        createNotification()
    }

    private fun createNotification() {
        val builder = Notification.Builder(applicationContext)
                .setContentIntent(PendingIntent.getActivity(this, 0, Intent(this, UniversalActivity::class.java), 0))
                .setLargeIcon(BitmapFactory.decodeResource(resources, R.mipmap.dk_doraemon))
                .setSmallIcon(R.mipmap.dk_doraemon)
                .setContentText("Dokit屏幕取色器前台服务")
                .setWhen(System.currentTimeMillis())
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            builder.setChannelId("notification_id")
            val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            val channel = NotificationChannel("notification_id", "notification_name", NotificationManager.IMPORTANCE_LOW)
            notificationManager.createNotificationChannel(channel)
        }

        val notification = builder.build()
        notification.defaults = Notification.DEFAULT_SOUND
        startForeground(110, notification)

    }

    companion object {

        fun bindService(context: Context, connection: ServiceConnection) {
            context.bindService(Intent(context, ScreenRecorderService::class.java), connection, BIND_AUTO_CREATE)
        }

        fun unbindService(context: Context, connection: ServiceConnection) {
            context.unbindService(connection)
        }
    }


    override fun onBind(intent: Intent?): IBinder? {
        return Binder()
    }
}